### DISCUSSION TITLE
Proposal to Simplify and Enhance main.go Structure

### DISCUSSION THREAD

**Discussion by almas-x** [goravel/goravel#548] [Ideas]
### Description:

The current main.go implementation manually manages service startup and shutdown logic, leading to repetitive code and a less modular structure. Below is an example of the current state:

```
package main

import (
	"os"
	"os/signal"
	"syscall"

	"github.com/goravel/framework/facades"

	"goravel/bootstrap"
)

func main() {
	// This bootstraps the framework and gets it ready for use.
	bootstrap.Boot()

	// Create a channel to listen for OS signals
	quit := make(chan os.Signal)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)

	// Start http server by facades.Route().
	go func() {
		if err := facades.Route().Run(); err != nil {
			facades.Log().Errorf("Route run error: %v", err)
		}
	}()

	// Listen for the OS signal
	go func() {
		<-quit
		if err := facades.Route().Shutdown(); err != nil {
			facades.Log().Errorf("Route Shutdown error: %v", err)
		}

		os.Exit(0)
	}()

	// Start grpc server by facades.Grpc().
	go func() {
		if err := facades.Grpc().Run(); err != nil {
			facades.Log().Errorf("Grpc run error: %v", err)
		}
	}()

	// Start queue server by facades.Queue().
	go func() {
		if err := facades.Queue().Worker().Run(); err != nil {
			facades.Log().Errorf("Queue run error: %v", err)
		}
	}()

	select {}
}
```

While functional, this approach has the following drawbacks:

1. **Repetitive Code**: Each service requires separate goroutines for startup and shutdown, duplicating signal handling logic.
2. **Tightly Coupled Logic**: Service lifecycle management is embedded directly in main.go, making it harder to maintain or extend.
3. **Limited Scalability**: Adding new services requires manual intervention, increasing the risk of errors.

### Proposed Solution:

To address these issues, I propose an optimized version of main.go that utilizes a ServiceManager for centralized and modular lifecycle management. Below is the refactored version:

```go
// main.go
package main

import (
	"github.com/goravel/framework/facades"

	"goravel/bootstrap"
)

func main() {
	// This bootstraps the framework and gets it ready for use.
	bootstrap.Boot()
	// Start http server by facades.Route().
	bootstrap.Start("Route", func() error {
		return facades.Route().Run()
	}, func() error {
		return facades.Route().Shutdown()
	})
	// Start grpc server by facades.Grpc().
	bootstrap.Start("Grpc", func() error {
		return facades.Grpc().Run()
	})
	// Start queue server by facades.Queue().
	bootstrap.Start("Queue", func() error {
		return facades.Queue().Worker().Run()
	})
	bootstrap.Wait()
}
```

```go
//  bootstrap/manager.go

package bootstrap

import (
	"context"
	"os"
	"os/signal"
	"sync"
	"syscall"

	"github.com/goravel/framework/facades"
)

var manager = &ServiceManager{}

func init() {
	manager.ctx, manager.cancel = context.WithCancel(context.Background())
        manager.wg.Add(1)
	// Create a channel to listen for OS signals
	go func() {
                defer manager.wg.Done()
		quit := make(chan os.Signal, 1)
		signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
		<-quit
		manager.Shutdown()
	}()
}

type ServiceManager struct {
	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

func Start(name string, starter func() error, stopper ...func() error) {
	manager.Start(name, starter, stopper...)
}
func (sm *ServiceManager) Start(name string, starter func() error, stopper ...func() error) {
	// start service
	go func() {
		if err := starter(); err != nil {
			facades.Log().Errorf("%s run error: %v", name, err)
		}
	}()
	// shutdown service
	if len(stopper) > 0 {
		sm.wg.Add(1)
		go func() {
			defer sm.wg.Done()
			<-sm.ctx.Done()
			if err := stopper[0](); err != nil {
				facades.Log().Errorf("%s shutdown error: %v", name, err)
			}
		}()
	}
}

func Wait() {
	manager.Wait()
}
func (sm *ServiceManager) Wait() {
	sm.wg.Wait()
}

func (sm *ServiceManager) Shutdown() {
	sm.cancel()
}
```

### Advantages of the Refactored Version:

1.  **Improved Readability**: The logic for starting and stopping services is abstracted, resulting in a cleaner main.go.
2.  **Centralized Lifecycle Management**: The ServiceManager handles signal listening, startup, and shutdown in a consistent manner.
3.  **Ease of Extension**: Adding new services only requires a new bootstrap.Start call with minimal code changes.
4.  **Better Scalability**: Modular design supports easier future enhancements.

### Call to Action:

I would like to gather feedback on whether this optimization is necessary and aligns with our project goals. Should we adopt this refactored structure, or are there other considerations or approaches we should evaluate?

> Your thoughts and suggestions are highly valued!

**Comment by hwbrzzl** [comment]
Great idea, yes, we need this. Two thoughts:

1. `ServiceManager` needs to be moved to goravel/framework, only add `Start`, `Wait` methods to goravel/goravel.
2. How about adding a new method: `bootstrap.Facades().Route()`, it's the same as: 

```
bootstrap.Start("Route", func() error {
  return facades.Route().Run()
}, func() error {
  return facades.Route().Shutdown()
})
```
